pytorch_lightning/trainer/trainer.py

```diff
@@ -481,7 +481,7 @@ def train(self):
                 # run train epoch
                 self.train_loop.run_training_epoch()
 
-                if self.max_steps and self.max_steps <= self.global_step:
+                if self.max_steps and self.max_steps <= self.global_step + 1:
```

Comment:
this can't be the fix, since this check here is at the end of each epoch. 
The fix must certainly be in the loop that runs over the training batches.

Reply:
once the training epoch returns after reaching max steps, we need to ensure we break out of the loop. this is what the fix does. you can see the logic is incorrect as it doesn't reflect the logic here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L582

Reply:
To summarise, the current PR fixes the issue by handling the assumption that with accumulate_grad_batches of 1 this line will always increment the global step: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L619

Now the check will prevent the step from going into an infinite loop as the global step is never incremented (when accumulate_grad_batches is greater than 1. 

When accumulate_grad_batches is greater than 1, this causes an infinite loop as the global step is never incremented and the checks are not in sync.

@awaelchli I'll double check the other test you also linked!

Reply:
ok, let's see. 
I merged master and uncommented my tests. Let's see if they work as intended :)

Reply:
Seems you are right, we're not in an infinite loop but we're missing a step. Will investigate further